accepted/0000-npm-option-handling.md

```diff
@@ -0,0 +1,190 @@
+and an even _more_ entrenched reliance on a god-object pattern, albeit one
+that is even more opinionated and harder to use than `npmconf` was.
+
+In one example where this has proven particularly challenging, `npm ci`
+does not get the full set of config values, and thus it does not properly
+respect config values that would be handled if the user ran `npm install`.
+Since _all_ package tree generation and reification will be handled by an
+external dependency in npm v7 (ie, `@npmcli/arborist`), this will become an
+even bigger issue.
+
+## Detailed Explanation
+
+In all of the CLI's dependencies, remove `figgy-pudding`, and replace with
+an options object that uses the canonical names for npm config values.
+
+In the npm CLI, create n options object with these canonically named
+fields, applying defaults and resolving aliases, as appropriate, using the
+`npm.config` object as its source of truth.
+
+### Canonical Naming
+
+Use the `camelCase` forms of all config fields.  For example, if the user
+runs `npm install --prefer-online`, then the config options object would
+contain `{ preferOnline: true }`.
+
+Thus, the npm CLI's dependencies should use `options.preferOnline` instead
+of `options['prefer-online']`.
+
+## Rationale and Alternatives
+
+It's worth noting the problem that figgy-pudding aims to solve and the value
+it currently provides, and ways that we will continue to solve that problem
+and gain that value.
+
+1. A figgy-pudding config object is a self-documenting list of the options
+   that a given module uses, with defaults, canonical name, and so on.
+2. Figgy-pudding allows a module to have a _different_ set of respected
+   configs than it provides to the modules deeper in the stack, because
+   they all proxy to the same underlying data object.
+3. Figgy-pudding prevents any accidental mutation of the options provided
+   to a module.
+4. Use of figgy-pudding avoids tight coupling to an `npmconf` object with
+   `get` and `set` methods.
+
+The response to these points is:
+
+1. It is better to document the list of options that a module uses _in its
+   documentation_, and make it clear in the code what their default values
+   are.  For example, instead of: 
+
+    ```js
+    const MyOptions = figgyPudding({
+      someFoo: { default: 'foo' },
+      myBar: {},
+    })
+
+    module.exports = (options) => {
+      options = MyOptions(options)
+      // ...options.someFoo
+    }
+    ```
+
+    We will now do:
+
+    ```js
+    module.exports = (options = {}) => {
+      const { someFoo = 'foo' } = options
+      // doSomething(someFoo)
+    }
+    ```
+
+    and document the `someFoo` field in the package `README.md`.
+
+2. When a module in the CLI dependency stack needs to call out to one lower
+   down (for example, arborist calling pacote, which in turn calls
+   npm-registry-fetch, which then calls make-fetch-happen, etc.), we can
+   just pass the options object to the deeper dep.
+
+    It's _fine_, really.
```

Comment:
Until someone forgets to clone and mutates the config object passed down. I've seen that anti-pattern far too many times inside npm-owned code (with no equivalent to eslint [no-param-reassign](https://eslint.org/docs/rules/no-param-reassign#props) enabled) to be especially confident of this.

_Especially_ when considering third-party consumers of said npm-owned libraries.

Reply:
Also, leaving aside the parameter hygiene issue, you now have to document (or at least be _aware_?) of config that isn't relevant in the context of a given module. I'm less confident folks will apply this "just pass around a bag of options" pattern 100% consistently across 100% of the stack without some help, either via static analysis, really really good memories, or some other (hopefully) automated process.

Reply:
So, figgy-pudding doesn't actually solve most of those problems.  Config hygiene has still been a huge problem.

There is no One Weird Trick for this.  It comes down to:

1. good docs that outline which configs are used
2. tests with 100% coverage
3. removing configs that are no longer relevant, and refactoring them when they don't make sense (see: only/also/production/optional/dev flags)
4. setting good defaults close to where they matter.

Third-party consumers of npm-owned libraries generally just passed in POJOs anyway, which would get figgy-ified.  The problem is that once a module is using figgy-pudding, everything _it_ uses must _also_ use figgy-pudding, and it has to be careful not to accidentally exclude an option that one of its deps will end up using, so every level of the stack effectively has to know about every other.  It's not encapsulation, it's the opposite of that.  And no one can debug its peculiar restrictions.

We can explore doing something clever with config management later, but for now, _nothing_, and just having good tests, is far better than the _something_ we have.

Reply:
> Until someone forgets to clone and mutates the config object passed down.

The recent move to 100% test coverage in all new code and for all dependencies has caught a ton of these types of problems, with far less hazardous side effects or bizarre restrictions imposed by figgy-pudding.

Reply:
Is there a reason not to Object.freeze it, and easily force the cloning?

Reply:
> Is there a reason not to Object.freeze it, and easily force the cloning?

I don't see why not.

Reply:
A frozen options object goes a long way toward mitigating my concerns (many of which, yes, no program can fix entirely).